bqskit/qis/pauliz.py

```diff
@@ -0,0 +1,251 @@
+"""This module implements the PauliMatrices class."""
+from __future__ import annotations
+
+import itertools as it
+from typing import Iterable
+from typing import Iterator
+from typing import overload
+from typing import Sequence
+from typing import TYPE_CHECKING
+
+import numpy as np
+import numpy.typing as npt
+
+from bqskit.utils.typing import is_integer
+from bqskit.utils.typing import is_numeric
+from bqskit.utils.typing import is_sequence
+
+if TYPE_CHECKING:
+    from bqskit.qis.unitary.unitary import RealVector
+
+
+class PauliZMatrices(Sequence[npt.NDArray[np.complex128]]):
+    """
+    The group of Pauli Z matrices.
+
+    A PauliZMatrices object represents the entire of set of Pauli Z matrices for
+    some number of qubits.
+    """
+
+    Z = np.array(
+        [
+            [1, 0],
+            [0, -1],
+        ], dtype=np.complex128,
+    )
+    """The Pauli Z Matrix."""
+
+    I = np.array(
+        [
+            [1, 0],
+            [0, 1],
+        ], dtype=np.complex128,
+    )
+    """The Identity Matrix."""
+
+    def __init__(self, num_qudits: int) -> None:
+        """
+        Construct the Pauli Z group for `num_qudits` number of qubits.
+
+        Args:
+            num_qudits (int): Power of the tensor product of the Pauli Z
+                group.
+
+        Raises:
+            ValueError: If `num_qudits` is less than or equal to 0.
+        """
+
+        if not is_integer(num_qudits):
+            raise TypeError(
+                'Expected integer for num_qudits, got %s.' %
+                type(num_qudits),
+            )
+
+        if num_qudits <= 0:
+            raise ValueError(
+                'Expected positive integer for num_qudits, got %s.' % type(
+                    num_qudits,
+                ),
+            )
+
+        self.num_qudits = num_qudits
+
+        if num_qudits == 1:
+            self.paulizs = [
+                PauliZMatrices.I,
+                PauliZMatrices.Z,
+            ]
+        else:
+            self.paulizs = []
+            matrices = it.product(
+                PauliZMatrices(
+                    num_qudits - 1,
+                ),
+                PauliZMatrices(1),
+            )
+            for pauliz_n_1, pauliz_1 in matrices:
+                self.paulizs.append(np.kron(pauliz_n_1, pauliz_1))
```

Comment:
Should we cache this calculation? I remember we did some caching for the Pauli matrices class.

Reply:
Can you point me towards this? I don't see it in the Pauli matrices class definition.

Reply:
nvm, we removed it a long time ago. I forget why, but there must have been a reason